File: SORTING/MergeSort.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>
void Merge(int A[], int l,int m,int r)
{
    int i,j,k,B[100000];
    i=l;
    j=m;
    k=l;
    while(i<m && j<=r)
    {
        if(A[i]<A[j])
        {
            B[k++]=A[i++];
        }
        else{
            B[k++]=A[j++];
        }
    }
    for( ;i<m;i++,k++)
    {
        B[k]=A[i];
    }
    for( ;j<=r;j++,k++)
    {
        B[k]=A[j];
    }
    for(k=l;k<=r;k++)
    {
        A[k]=B[k];
    }

}
void MergeSort(int A[],int l,int r)
{
    int m;
    if(l<r)
    {
        m=(l+r)/2;
        MergeSort(A,l,m);
        MergeSort(A,m+1,r);
        Merge(A,l,m+1,r);
    }
}

void display(int A[],int n)
{
    int i;
    for(i=0;i<n;i++)
    {
        printf("%d ",A[i]);
    }
    printf("\n");

}

int main()
{
    int A[100000],n,i;
    clock_t start,end,diff;
    printf("Enter n:"); 
    scanf("%d",&n);
    for(i=0;i<n;i++)
    {
        A[i]=rand();
    }
    // display(A,n);
    start=clock();
    MergeSort(A,0,n-1);
    end=clock();
    printf("The sorted numbers are:\n");
    // display(A,n);
    diff=end-start;
    printf("The time taken is %f seconds\n",(float)diff/CLOCKS_PER_SEC);
    return 0;
}

File: SORTING/SortingInPython/mergeSort.py
import random
import time

def Merge(A, l, m, r):
    B = [0] * 100000  # Temporary array
    j = m
    k = l
    while i < m and j <= r:
        if A[i] < A[j]:
            B[k] = A[i]
            i += 1
        else:
            B[k] = A[j]
            j += 1
        k += 1
    while i < m:
        B[k] = A[i]
        i += 1
        k += 1
    while j <= r:
        B[k] = A[j]
        j += 1
        k += 1
    for k in range(l, r + 1):
        A[k] = B[k]

def MergeSort(A, l, r):
    if l < r:
        m = (l + r) // 2
        MergeSort(A, l, m)
        MergeSort(A, m + 1, r)
        Merge(A, l, m + 1, r)

def display(A, n):
    for i in range(n):
        print(A[i], end=" ")
    print()

A = [0] * 100000
print("Enter n:", end=" ")
n = int(input())
for i in range(n):
    A[i] = random.randint(0, 100000)
start = time.time()
MergeSort(A, 0, n - 1)
end = time.time()
print("The sorted numbers are:")
print(f"The time taken is {end - start} seconds")
